### PR TITLE
Avoid cycle in tuple explicit interface implementation

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/60961.

The cycle was roughly like this:

ForceComplete ValueTuple<T1, T2, ...>
-> CheckBase
-> Bind (T1, T2...)
-> CreateTuple -> GetWellKnownMemberInType(TRest)
-> GetMembers of ValueTuple<T1, T2, ...> with name "TRest"
-> AddNonTypeMembers -> indexer (explicit implementation)
-> SourcePropertySymbol.Create
-> ExplicitInterfaceHelpers.GetMemberNameAndInterfaceSymbol
-> BindType ITuple
-> ...bind components of ITuple's namespace...
-> BindNamespaceOrTypeSymbol System in ValueTuple
-> LookupSymbolsSimpleName "System" in ValueTuple
-> LookupMembersInType "System" in ValueTuple
-> LookupMembersInClass (here basesBeingResolved is checked)
-> GetNextBaseTypeNoUseSiteDiagnostics
-> MakeDeclaredBases (here basesBeingResolved is populated)
-> BindType IEquatable<ValueTuple<...>>
-> BindTupleType -> CreateTuple (here basesBeingResolved is lost)
-> cycle

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs [hunk 1]
```diff
@@ -214,5 +215,37 @@ public void TestMissingWellKnownMembersForValueTuple()
...
+                //     public struct ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest> : System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)> where TRest : struct
+                Diagnostic(ErrorCode.ERR_UnimplementedInterfaceMember, "System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>").WithArguments("System.ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>", "System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>.Equals((T1, T2, T3, T4, T5, T6, T7, TRest))").WithLocation(4, 67),
+                // (14,17): error CS0540: 'ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>.ITuple.this[int]': containing type does not implement interface 'ITuple'
+                //         object? System.Runtime.CompilerServices.ITuple.this[int index] => throw null!;
+                Diagnostic(ErrorCode.ERR_ClassDoesntImplementInterface, "System.Runtime.CompilerServices.ITuple").WithArguments("System.ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>.System.Runtime.CompilerServices.ITuple.this[int]", "System.Runtime.CompilerServices.ITuple").WithLocation(14, 17));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:249]: >} [](http://example.com/codeflow?start=8&length=1)

Should we also test a valid implementation?
```csharp
namespace System
{
    public struct ValueTuple<T1>
    {
        public T1 Item1;
        public ValueTuple(T1 item1) { }
    }
    public struct ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest> :
        System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>,
        System.Runtime.CompilerServices.ITuple
        where TRest : struct
    {
        public T1 Item1;
        public T2 Item2;
        public T3 Item3;
        public T4 Item4;
        public T5 Item5;
        public T6 Item6;
        public T7 Item7;
        public TRest Rest;
        object? System.Runtime.CompilerServices.ITuple.this[int index] { get { throw null; } }
        bool System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>.Equals((T1, T2, T3, T4, T5, T6, T7, TRest) other) => false;
    }
}
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs [hunk 2]
```diff
@@ -214,5 +215,37 @@ public void TestMissingWellKnownMembersForValueTuple()
             Assert.Null(comp.GetWellKnownTypeMember(WellKnownMember.System_ValueTuple_TRest__Item7));
             Assert.Null(comp.GetWellKnownTypeMember(WellKnownMember.System_ValueTuple_TRest__Rest));
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/60961")]
+        public void ExplicitInterfaceImplementation()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:220]: >ExplicitInterfaceImplementation [](http://example.com/codeflow?start=20&length=31)

Is it possible to add similar test for VB?

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs [hunk 3]
```diff
@@ -52,7 +53,8 @@ internal abstract partial class NamedTypeSymbol
 
             NamedTypeSymbol underlyingType = getTupleUnderlyingType(elementTypesWithAnnotations, syntax, compilation, diagnostics);
 
-            if (numElements >= ValueTupleRestPosition && diagnostics != null && !underlyingType.IsErrorType())
+            if (numElements >= ValueTupleRestPosition && diagnostics != null && !underlyingType.IsErrorType() &&
+                basesBeingResolved?.ContainsReference(underlyingType.OriginalDefinition) != true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:57]: >basesBeingResolved?.ContainsReference(underlyingType.OriginalDefinition) != true [](http://example.com/codeflow?start=16&length=80)

While I understand that this change prevents a cycle, the logic looks suspicious. Usually, a `basesBeingResolved` check leads to some form of a cancelation, an error is reported, or items in the list are getting skipped in some processing. This doesn't look like that, and it doesn't have the same effect. 
Also, I think that we are not fixing the root cause problem for the cycle, which, in my opinion, is the fact that we are trying to bind something while creating symbols for declared members in type. There were other instances of similar "too early" binding, but we got rid of them. The indexer name is probably the last one that we didn't address. Since we finally have a motivating test scenario, I think we should get rid of the binding by changing the naming scheme for indexers. There must be an alternative to using ```explicitInterfaceTypeOpt.ToDisplayString(SymbolDisplayFormat.ExplicitInterfaceImplementationFormat)``` for building a unique name. For example, we could simply use explicit interface specifier followed by the name identifier as it is spelled out in source, combined with the containing partial type declaration index.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs [hunk 4]
```diff
@@ -52,12 +52,6 @@ internal abstract partial class NamedTypeSymbol
 
             NamedTypeSymbol underlyingType = getTupleUnderlyingType(elementTypesWithAnnotations, syntax, compilation, diagnostics);
 
-            if (numElements >= ValueTupleRestPosition && diagnostics != null && !underlyingType.IsErrorType())
-            {
-                WellKnownMember wellKnownTupleRest = GetTupleTypeMember(ValueTupleRestPosition, ValueTupleRestPosition);
-                _ = GetWellKnownMemberInType(underlyingType.OriginalDefinition, wellKnownTupleRest, diagnostics, syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:58]: >_ = GetWellKnownMemberInType(underlyingType.OriginalDefinition, wellKnownTupleRest, diagnostics, syntax); [](http://example.com/codeflow?start=16&length=105)

Are we going to "intercept" the errors, that possibly used to be reported here, some where else?

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:58]: It seems to me that's not necessary - whoever is using the `Rest` member will report those errors if it's missing.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:58]: > It seems to me that's not necessary ...

It sounds like the answer is "yes", the error will be reported by someone else. At least this is the hope. How confident you are and what is the basis for that level of confidence?

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:58]: > It sounds like the answer is "yes"

Yes, sorry, I misunderstood the question.

> How confident are you, and what is the basis for that level of confidence?

If someone wants to access the `Rest` member, they should do that via `WellKnownMember` enum in a way that reports diagnostics if that member is missing. I've looked through usages of `NamedTypeSymbol.GetTupleTypeMember` and they seem to be doing that where necessary.

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs (hunk 3 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: After offline discussion, filed an issue to track this: https://github.com/dotnet/roslyn/issues/70901.

File: src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs [hunk 5]
```diff
@@ -268,7 +268,7 @@ public class SymbolDisplayFormat
                 globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.OmittedAsContaining,
                 typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
                 genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
-                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers,
+                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers | SymbolDisplayMiscellaneousOptions.ExpandValueTuple,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:271]: > | SymbolDisplayMiscellaneousOptions.ExpandValueTuple [](http://example.com/codeflow?start=96&length=53)

Why is this change necessary?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 6]
```diff
@@ -20498,10 +20498,12 @@ public class C4 : I2, I3
 
             void assertExplicitInterfaceImplementations(INamedTypeSymbol c)
             {
-                var cMabImplementations = ((IMethodSymbol)c.GetMember("I1<(System.Int32a,System.Int32b)>.M")).ExplicitInterfaceImplementations;
+                var members = c.GetMembers("I1<System.ValueTuple<System.Int32,System.Int32>>.M");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:20501]: >I1<System.ValueTuple<System.Int32,System.Int32>> [](http://example.com/codeflow?start=44&length=48)

I am not sure if change in the member name is desirable.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs [hunk 7]
```diff
@@ -214,5 +215,391 @@ public void TestMissingWellKnownMembersForValueTuple()
...
+                        public TRest Rest;
+                        int System.Runtime.CompilerServices.ITuple.Length => throw null;
+                        public object this[int index] => throw null;
+                        bool System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>.Equals((T1, T2, T3, T4, T5, T6, T7, TRest) other) => false;
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:360]: >VerifyDiagnostics [](http://example.com/codeflow?start=17&length=17)

VerifyEmitDiagnostics?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs [hunk 8]
```diff
@@ -214,5 +215,391 @@ public void TestMissingWellKnownMembersForValueTuple()
...
+                            object this[int index] { get; }
+                        }
+                    }
+
+                    public delegate void D();
+                }
+                """;
+
+            var comp = CreateEmptyCompilation(source);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:422]: >VerifyDiagnostics [](http://example.com/codeflow?start=17&length=17)

VerifyEmitDiagnostics?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs [hunk 9]
```diff
@@ -214,5 +215,391 @@ public void TestMissingWellKnownMembersForValueTuple()
...
+                            object this[int index] { get; }
+                            T this[T self, int index] { get; }
+                        }
+                    }
+                }
+                """;
+
+            var comp = CreateEmptyCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:603]: > } [](http://example.com/codeflow?start=7&length=2)

It looks like there are no tests with tuple types used in the signature of the implemented member (I recommended to add them in our offline conversation)

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 10]
```diff
@@ -23371,6 +23371,121 @@ True
...
+                Return Nothing
+            End Get
+        End Property
+        Function IEquatable_Equals(other As (T1, T2, T3, T4, T5, T6, T7, TRest)) As Boolean Implements System.IEquatable(Of (T1, T2, T3, T4, T5, T6, T7, TRest)).Equals
+            Return False
+        End Function
+    End Structure
+End Namespace"
+            Dim comp = CreateCompilation(source, targetFramework:=TargetFramework.NetCoreApp)
+            comp.AssertTheseDiagnostics()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:23487]: >AssertTheseDiagnostics [](http://example.com/codeflow?start=17&length=22)

Verify emit diagnostics?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs (hunk 9 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:603]: I thought I added them. For example, `ExplicitInterfaceImplementation_Indexer` has `bool System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>.Equals((T1, T2, T3, T4, T5, T6, T7, TRest) other) => false;` - the implemented member has tuple types in signature. Or did you mean like non-generic tuple types?

File: src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs (hunk 5 above)
Comment (jjonescz) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:271]: To avoid a cycle when completing tuple members and some member has tuple in signature (e.g., `bool System.IEquatable<(T1, T2, T3, T4, T5, T6, T7, TRest)>.Equals((T1, T2, T3, T4, T5, T6, T7, TRest) other)` in tests), it will go through `SymbolDisplayVisitor` to compute the explicitly implemented member name which will in turn bind the tuple type and continue the cycle.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 6 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:20501]: Originally you proposed to change all explicitly implemented member names, so why is this one a problem?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs (hunk 9 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ValueTupleTests.cs:603]: > . Or did you mean like non-generic tuple types?

I meant tests that do not refer to tuples anywhere, but in the signature of the implemented memeber.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:20501]: > Originally you proposed to change all explicitly implemented member names, so why is this one a problem?

I do not know the reason behind the change. Hence the concern.

File: src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:271]: What is the source of the cycle in this case?

Comment (jjonescz) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs:271]: If you mean how the cycle happens, I tried to describe that in the comment above, but here it is in more detail:
1. compilation is forced to complete
2. tuple members are being created
3. explicitly implemented members are creating their name
4. they use `SymbolDisplayVisitor` for the interface type
5. the interface type contains the tuple, so the visitor queries the tuple's members to display it
6. go to step 2

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Review by cston — APPROVED** [review]
